### DISCUSSION TITLE
How to cancel a task correctly?

### DISCUSSION THREAD

**Discussion by Alex0Klain** [pointfreeco/swift-composable-architecture#2307] [Q&A]
Hello!
I don't quite understand what's wrong here. I want to cancel the previous `.buildCommand` action and leave the last one. `.someAction` can be called a lot of times and you need to leave sending only the last action. But actually when called `.buildCommand`, if you call `.someAction` again, then last `.buildCommand` will also be canceled for some reason. If you don't call `.someAction` again during this `.buildCommand` call, then everything will work correctly

**Reducer:**
```
   case .someAction:
    return EffectTask.concatenate(
      EffectTask.merge(
        EffectTask(value: .action1),
        EffectTask(value: .action2),
        EffectTask(value: .action3)
      ),
      EffectTask.concatenate(
        .cancel(id: CancelID.changeTemperature),
        .task {
          try? await Task.sleep(nanoseconds: 2_000_000_000)
          return .buildCommand
        }
        .cancellable(
          id: CancelID.changeTemperature
        )
      )
    )
```

**Comment by mbrandonw** [comment]
The problem is that you are using `try? await` for the sleep. When that effect is cancelled it causes `Task.sleep` to throw, but that error is discarded with `try?` and so the `.buildCommand` is emitted by the effect. I think you just want `try await`.

**Reply by Alex0Klain** [reply]
I changed to `try await`, but unfortunately it didn't solve the problem. So if you call `.someAction` during these `2 seconds` while `Task.sleep` is active, then `.buildCommand` will not be sent

**Reply by mbrandonw** [reply]
I'm not sure I understand your description of what is happening versus what you expect to happen. Can you extra the code to a project that one can run and see the behavior?

**Reply by Alex0Klain** [reply]
I expect that clicking on `.someAction` will `cancel` the last one `.buildCommand` and go from the current `.someAction`. But in fact, if immediately after sending `.buildCommand` _(within 2 seconds)_ click again on `.someAction`, then the effect of the past and current `.buildCommand` will be `canceled`.
Unfortunately, I cannot provide the project because it is under `NDA` :(

**Reply by mbrandonw** [reply]
Sorry I am still not understanding. I don't need access to the full project, but if you want help I think you need to show some concrete code. Try building something from scratch and share it.
